### ISSUE TITLE
(t) System-Services, Storage-Replication-Send_Receive, and Tasks pages fail

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2370]
(t) = testing branch. We have an outstanding page failure post our recent multi Django version update in #2365.
```
Unknown client error doing a GET to /api/sm/services/?page=1&format=json&page_size=32000&count= 
```
no entries in nginx.error and the expected 404 in nginx.access (includes prior successful '200' disks page access):
```
 - - [18/Apr/2022:19:15:30 +0100] "GET /api/disks?page=1&format=json&page_size=32000&count= HTTP/1.1" 200 3028 "https://rleap15-3.lan/" "Mozilla/5.0 (X11; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0"
 - - [18/Apr/2022:19:16:05 +0100] "GET /api/sm/services/?page=1&format=json&page_size=32000&count= HTTP/1.1" 404 1939 "https://rleap15-3.lan/" "Mozilla/5.0 (X11; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0"
```

**Comment by FroggyFlox** [comment]
@phillxnet, I had a look at that one, and will share my notes below hoping it'll provide some clues and maybe not distract too much from the root cause of the issue here.
Note that this "exploration" was done using my branch with #2378 applied so that I can easily restart/rebuild without worrying about the database migrations.

As we couldn't really find anything in the system/rockstor logs besides what you listed, I tried to use Firefox developer tools to have a better idea of the request that sent and the response received.

# Current testing branch
# Testing
## Headers
GET
scheme: https
host: buildvm
filename: /api/sm/services/
page: 1
format: json
page_size: 32000
count: 
Address: 192.168.122.217:443
Status: 404 Not Found
Version HTTP/1.1
Transferred 2.08 KB (15.76 KB size)
Referrer Policy strict-origin-when-cross-origin

### Response Headers
HTTP/1.1 404 Not Found
Date: Wed, 15 Jun 2022 12:29:50 GMT
Content-Type: text/html
Transfer-Encoding: chunked
Connection: keep-alive
Keep-Alive: timeout=20
Server: gunicorn/19.7.1
Content-Encoding: gzip

### Request Headers
GET /api/sm/services/?page=1&format=json&page_size=32000&count= HTTP/1.1
Host: buildvm
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:101.0) Gecko/20100101 Firefox/101.0
Accept: application/json, text/javascript, */*; q=0.01
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
X-Requested-With: XMLHttpRequest
DNT: 1
Connection: keep-alive
Referer: https://buildvm/
Cookie: csrftoken=vKI2xm9A7c332waX3HCXGgk586orzfxhX2htfqv0rkonEwAkQ5W8QqUjZmQEh6Fr; sessionid=ku5pjwc790127pr1afkj3jmpgpsda88l
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin

## Response

Page not found (404)
Request Method: 	GET
Request URL: 	https://buildvm/api/sm/services/?page=1&format=json&page_size=32000&count=

Using the URLconf defined in urls, Django tried these URL patterns, in this order:

    ^$ [name='home']
    ^login_page$ [name='login_page']
    ^login_submit$ [name='login_submit']
    ^logout_user$ [name='logout_user']
    ^home$ [name='home']
    ^setup_user$
    ^site_media/(?P<path>.*)$
    ^css/(?P<path>.*)$
    ^js/(?P<path>.*)$
    ^img/(?P<path>.*)$
    ^o/
    ^api/login
    ^api/appliances$
    ^api/appliances/(?P<appid>\d+)$
    ^api/commands/
    ^api/disks$
    ^api/disks/
    ^api/network$
    ^api/network/
    ^api/pools$ [name='pool-view']
    ^api/pools/
    ^api/shares$ [name='share-view']
    ^api/shares/
    ^api/snapshots
    ^api/users$
    ^api/users/(?P<username>[A-Za-z][-a-zA-Z0-9_]*$)$
    ^api/groups$
    ^api/groups/(?P<groupname>[A-Za-z][-a-zA-Z0-9_]*$)$
    ^api/nfs-exports$
    ^api/nfs-exports/(?P<export_id>\d+)$
    ^api/adv-nfs-exports$
    ^api/samba$
    ^api/samba/(?P<smb_id>\d+)$
    ^api/sftp$
    ^api/sftp/(?P<id>\d+)$
    ^api/dashboardconfig$
    ^api/oauth_app$
    ^api/oauth_app/(?P<id>\d+)$
    ^api/sm/services$
    ^api/sm/services/ ^nis$
    ^api/sm/services/ ^nis/(?P<command>config|start|stop)$
    ^api/sm/services/ ^smb$
    ^api/sm/services/ ^smb/(?P<command>config|start|stop)$
    ^api/sm/services/ ^nfs$
    ^api/sm/services/ ^nfs/(?P<command>config|start|stop)$
    ^api/sm/services/ ^ntpd$
    ^api/sm/services/ ^ntpd/(?P<command>config|start|stop)$
    ^api/sm/services/ ^active-directory$
    ^api/sm/services/ ^active-directory/(?P<command>config|start|stop)$
    ^api/sm/services/ ^ldap$
    ^api/sm/services/ ^ldap/(?P<command>config|start|stop)$
    ^api/sm/services/ ^sftp$
    ^api/sm/services/ ^sftp/(?P<command>config|start|stop)$
    ^api/sm/services/ ^replication$
    ^api/sm/services/ ^replication/(?P<command>config|start|stop)$
    ^api/sm/services/ ^task-scheduler$
    ^api/sm/services/ ^task-scheduler/(?P<command>config|start|stop)$
    ^api/sm/services/ ^data-collector$
    ^api/sm/services/ ^data-collector/(?P<command>config|start|stop)$
    ^api/sm/services/ ^service-monitor$
    ^api/sm/services/ ^service-monitor/(?P<command>config|start|stop)$
    ^api/sm/services/ ^snmpd$
    ^api/sm/services/ ^snmpd/(?P<command>config|start|stop)$
    ^api/sm/services/ ^docker$
    ^api/sm/services/ ^docker/(?P<command>config|start|stop)$
    ^api/sm/services/ ^smartd$
    ^api/sm/services/ ^smartd/(?P<command>config|start|stop)$
    ^api/sm/services/ ^nut$
    ^api/sm/services/ ^nut/(?P<command>config|start|stop)$
    ^api/sm/services/ ^ztask-daemon$
    ^api/sm/services/ ^ztask-daemon/(?P<command>config|start|stop)$
    ^api/sm/services/ ^rockstor-bootstrap$
    ^api/sm/services/ ^rockstor-bootstrap/(?P<command>config|start|stop)$
    ^api/sm/services/ ^shellinaboxd$
    ^api/sm/services/ ^shellinaboxd/(?P<command>config|start|stop)$
    ^api/sm/services/ ^rockstor$
    ^api/sm/services/ ^rockstor/(?P<command>config|start|stop)$
    ^api/sm/sprobes$ [name='probe-view']
    ^api/sm/sprobes/
    ^api/sm/tasks$
    ^api/sm/tasks/
    ^api/sm/replicas$ [name='replica-view']
    ^api/sm/replicas/
    ^api/certificate
    ^api/rockons$
    ^api/rockons/
    ^api/config-backup$
    ^api/config-backup/(?P<backup_id>\d+)$
    ^api/config-backup/file-upload$
    ^api/email$
    ^api/email/(?P<command>.*)$
    ^api/pincardmanager/(?P<command>create|reset)/(?P<user>\w+)$
    ^api/update-subscriptions$
    ^api/update-subscriptions/

The current path, api/sm/services/, didn't match any of these.

You're seeing this error because you have DEBUG = True in your Django settings file. Change that to False, and Django will display a standard 404 page.

# STABLE
## Headers
GET
scheme: https
host: 192.168.122.232
filename: /api/sm/services/
page: 1
format: json
page_size: 32000
count: 
Address: 192.168.122.232:443
Status: 200 OK
Version HTTP/1.1
Transferred 2.52 KB (2.28 KB size)
Referrer Policy strict-origin-when-cross-origin

### Response Headers
HTTP/1.1 200 OK
Date: Wed, 15 Jun 2022 12:38:38 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive
Keep-Alive: timeout=20
Server: gunicorn/19.7.1
Vary: Accept, Cookie
Allow: GET, POST, HEAD, OPTIONS

### Request Headers
GET /api/sm/services/?page=1&format=json&page_size=32000&count= HTTP/1.1
Host: 192.168.122.232
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:101.0) Gecko/20100101 Firefox/101.0
Accept: application/json, text/javascript, */*; q=0.01
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
X-Requested-With: XMLHttpRequest
DNT: 1
Connection: keep-alive
Referer: https://192.168.122.232/
Cookie: csrftoken=GGqkuDMtzBXpwrWNFhIYae3ToCKSJrcY; sessionid=v7evknpb7wwxif57did0c6idz60wpq1o
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin

**Comment by FroggyFlox** [comment]
It thus seem to confirm that the request sent is the same in both branches, at least. What is curious, though, was the following line in the TESTING debug:
> The current path, `api/sm/services/`, didn't match any of these.

with the search path listing:
> `^api/sm/services$`

I tried to compare this to another page, like the Samba export, where we have the filename in the request as (Firefox developer tool output):
> filename: `/api/samba`

Note the lack of trailing slash there, and its url in our django being set as:
```python
    url(r"^api/samba$", SambaListView.as_view()),
```
Interestingly, our services url is set the same way, with the absence of trailing slash:
```python
    url(r"^api/sm/services$", BaseServiceView.as_view()),
```

I thus wonder if we erroneously have a trailing slash somewhere for our services page.

**Comment by FroggyFlox** [comment]
Interestingly, we do have:
https://github.com/rockstor/rockstor-core/blob/57998ec6980510f5efef6e3adb2bfb7b2b51d03c/src/rockstor/storageadmin/static/storageadmin/js/models/models.js#L312-L320

To continue our comparison with the Samba page, `models.js` contains:
https://github.com/rockstor/rockstor-core/blob/57998ec6980510f5efef6e3adb2bfb7b2b51d03c/src/rockstor/storageadmin/static/storageadmin/js/models/models.js#L302-L310

**Comment by FroggyFlox** [comment]
Note that all of the above is also true for the *Storage* > *Replication pages*... Removing the trailing slash to `/api/sm/replicas/` in `models.js` makes these pages accessible again.

**Comment by phillxnet** [comment]
Thanks @FroggyFlox this one had been bothered me quite some. I also got as far as the trailing "/" with working counterpart but failed to get further. Nice run-down.
I've now popped your prior pending testing pr in to ease further testing here.

**Comment by FroggyFlox** [comment]
@phillxnet , thanks a lot for merging the other PR. I've now rebased on upstream/testing and tested the following:
- can visit the System > Services page without error
- can visit the Storage > Replication pages without error
- can configure and toggle ON/OFF services (tested Samba and Rock-ons).

It thus seems to be ok at first glance, but I'm entirely familiar with how URLs are set in Django and I know you had to make a lot of changes in them during the Django upgrade so I think I'll rely on your knowledge here to make sure that modifying the `models.js` bits is the proper way to go whenever you can get to it. One could also modify the URLs to have that trailing slash, for instance... The fact that our current syntax was working on previous Django versions but not the new one really highlights how important the management of URLs by Django is in this case, and thus how I'd rather rely on your experience with the topic here (at your own convenience, of course).

Cheers,

**Comment by phillxnet** [comment]
@FroggyFlox Cheers.
Re:
> ... but I'm not entirely familiar with how URLs are set in Django

Yes me too, they did change the way paths were defined a little in the jump made so far in testing and they will all need to be changed again again when we get to where @Hooverdan96 has paved the way re the even newer Django in:
(t) Experimental Django update #2372
and it's new url definition type of re_path(). So we need in time to move these all over again with the next major Django update! But all in good time.

I think what has lead us to this issue, now sorted apparently by yourself, is a change of defaults regarding trailing "/" in Django versions transitioned in our already large version jumps done to date. Note that I played a few times with the setting of:

https://github.com/rockstor/rockstor-core/blob/6e66f3b5210f0c5b94022704a04471f13a563dbe/conf/settings.conf.in#L155-L158

And in attempting to keep stuff working and honour our existing prior behaviour I went with the setting above. But it may well be we have to re-visit this in time. I know it was a right headache at the time to know which way to turn and I plumped for that setting to get us along some. The auto trailing setting just seemed to break more of our current arrangement than sticking to what had already worked. I suspect we will have to revisit in time but for now returning a much function as possible is the key factor as I see it.

But at least now we have our url definitions more centralised so any changes/inconsistencies should be easier to see.

Again nice find and thanks for tracking that one down.

I say, as usual, that function trumps form and here we have a return of function and if we pop in it's own pr we have a focused record of the change, just in case it turns out we took a wrong turning. I'll prepare a pr with attribution to this thread soon so we can get these two remaining Web-IU areas back into play. Not sure it's the right tack but easier to fix when we see it working incorrectly than not a tall I think. Plus we need to get the code out in the testing channel as it's been far too long here only in the GitHub testing branch and not out in field testing via the testing channel.

**Comment by Hooverdan96** [comment]
Is this entry in the model.js also relevant w.r.t. to the trailing slash you uncovered? I did not test whether it is causing problems.

https://github.com/rockstor/rockstor-core/blob/57998ec6980510f5efef6e3adb2bfb7b2b51d03c/src/rockstor/storageadmin/static/storageadmin/js/models/models.js#L607-L611

**Comment by phillxnet** [comment]
@Hooverdan96 Hello there.
Re:

> Is this entry in the models.js also relevant w.r.t. to the trailing slash you uncovered?

Potentially. I'm really not that sure. But once we have the majority of stuff working we can take it from there. I'll hopefully be looking at this issue very shortly so it may very well be a repeat. I still a little way out on this one but edging closer to having a proper look. I've not looked at tasks at all in the testing channel yet. I'll give them a go once I've popped in a pr with @FroggyFlox 's findings and we can then see if the same treatment is needed there.

Watch this space. If so I'll add it to the meta issue of:
https://github.com/rockstor/rockstor-core/issues/2362
and to this issues subject.

Thanks for taking a look here. We may just be building up more technical debt but we have so much larger issues to address in relatively short shrift that I'm keen to be getting on with the Python 2 -> 3 move of:
"move to python 3" #1877
That we are lining up to with all these post Django update fixes.

**Comment by phillxnet** [comment]
@Hooverdan96 Oh year!!
Re:

> Is this entry in the models.js also relevant w.r.t. to the trailing slash you uncovered?

and:

> we can then see if the same treatment is needed there.

```
192.168.2.138 - - [17/Jun/2022:19:10:59 +0100] "GET /api/sm/tasks/?page=1&format=json&page_size=32000&count= HTTP/1.1" 404 1690 "https://rleap15-3.lan/" "Mozilla/5.0 (X11; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0"
```

@FroggyFlox proposed fix/work-around applied similarly:

```
192.168.2.138 - - [17/Jun/2022:19:45:14 +0100] "GET /api/sm/tasks?page=1&format=json&page_size=32000&count= HTTP/1.1" 200 63 "https://rleap15-3.lan/" "Mozilla/5.0 (X11; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0"
192.168.2.138 - - [17/Jun/2022:19:45:14 +0100] "GET /api/sm/tasks/log?page=1&format=json&page_size=32000&count= HTTP/1.1" 200 63 "https://rleap15-3.lan/" "Mozilla/5.0 (X11; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0"
```
Although if we then try to set a task we end up running into another issue re:

```
192.168.2.138 - - [17/Jun/2022:19:54:28 +0100] "POST /api/sm/tasks/ HTTP/1.1" 404 1656 "https://rleap15-3.lan/" "Mozilla/5.0 (X11; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0"
```
However at least we now have a reproducer for that from the Web-UI. And we have an entire other layer of issues that will be fixed when we address just opened: #2379 .

So again, given we need to address things in layers and bit-by-bit I think I'm going to include this tasks thing also. Still having a final think on it thought.

**Comment by FroggyFlox** [comment]
Nice catch, @Hooverdan96 , and thanks @phillxnet for testing it out!


> Although if we then try to set a task we end up running into another issue re:
> 
> ```
> 192.168.2.138 - - [17/Jun/2022:19:54:28 +0100] "POST /api/sm/tasks/ HTTP/1.1" 404 1656 "https://rleap15-3.lan/" "Mozilla/5.0 (X11; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0"
> ```
> 
> However at least we now have a reproducer for that from the Web-UI.

I suspect we need to fix the following as well:

https://github.com/rockstor/rockstor-core/blob/81a0d5f5e0a6dfe5a922199828f66eeab0253e65/src/rockstor/storageadmin/static/storageadmin/js/views/add_scheduled_task.js#L230-L232

**Comment by phillxnet** [comment]
@FroggyFlox I've literally just tested that myself (before looking here by the way). And yes. that does fix things. Now looking at the same for replications. I thing we are getting there now. I'll post pr with these fixes next. Once we have the basic functionality.

Need to do more setup for testing pre and post trailing for type POST on replication but likely has the exact same issue. Looked at a few others and I think they may on the hole be better behaved.

I should really keep this window open.
```
192.168.2.138 - - [17/Jun/2022:21:53:51 +0100] "GET /api/sm/tasks/1 HTTP/1.1" 200 236 "https://rleap15-3.lan/" "Mozilla/5.0 (X11; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0"
```
task created via Web-UI (@FroggyFlox I think tecnically you may have beat me to it :) ).

**Comment by phillxnet** [comment]
OK, so yes, we have the same issue with replication. Testing proposed change here also:
```
Unknown client error doing a POST to /api/sm/replicas/ 
```

**Comment by phillxnet** [comment]
@FroggyFlox & @Hooverdan96 
OK so that seems to have learnt it.
![add-replication](https://user-images.githubusercontent.com/2521585/174403105-78431cf0-e392-4fb9-bd2a-33b86889c203.png)

We may well have some more hard-coded, now incompatible url railing "/" in the same area but at least we now have the add task and add replication sent sorted.  I'll pop in a pr for review as there was an additional related fix that looks to also be required. Plus easier test once in pr.

**Comment by phillxnet** [comment]
Fixed by #2380
